Source/ablastr/fields/MultiFabRegister.H

```diff
@@ -0,0 +1,713 @@
+     *
+     * This class is used to own the lifetime of an amrex::MultiFab and to store
+     * associated information around it regarding unique naming, scalar/vector/tensor
+     * properties, aliasing, load balancing, etc.
+     */
+    struct MultiFabOwner
+    {
+        // TODO: also add iMultiFab via std::variant
+
+        /** owned (i)MultiFab */
+        amrex::MultiFab m_mf;
+
+        /** Components (base vector directions) of this (i)MultiFab */
+        std::optional<Direction> m_dir = std::nullopt;
+
+        /** the MR level of this (i)MultiFab */
+        int m_level = 0;
+
+        /** remake distribution map on load balance, @see amrex::AmrCore::RemakeLevel */
+        bool m_remake = true;
+
+        /** redistribute on @see amrex::AmrCore::RemakeLevel */
+        bool m_redistribute_on_remake = true;
+
+        /** if m_mf is a non-owning alias, this string tracks the name of the owner */
+        std::string m_owner;
+
+        /** Is this part of a vector/tensor? */
+        AMREX_INLINE
+        bool
+        is_vector () const { return m_dir.has_value(); }
+
+        /** Is this an alias MultiFab?
+         *
+         * If yes, that means we do not own the memory.
+         */
+        AMREX_INLINE
+        bool
+        is_alias () const { return !m_owner.empty(); }
+    };
+
+    /** This is a register of fields aka amrex::(i)MultiFabs.
+     *
+     * This is owned by a simulation instance. All used fields should be registered here.
+     * Internally, this contains @see MultiFabOwner values.
+     */
+    struct MultiFabRegister
+    {
+        // Avoid accidental copies when passing to member functions
+        MultiFabRegister() = default;
+        MultiFabRegister(MultiFabRegister const &) = delete;
+        MultiFabRegister(MultiFabRegister&&) = delete;
+        MultiFabRegister& operator=(MultiFabRegister const &) = delete;
+        MultiFabRegister& operator=(MultiFabRegister&&) = delete;
+        ~MultiFabRegister() = default;
+
+        /** Allocate and optionally initialize a MultiFab (field)
+         *
+         * This registers a new MultiFab under a unique name, allocates it and
+         * optionally assigns it an initial value.
+         *
+         * @param name a unique name for this field
+         * @param level the MR level to represent
+         * @param ba the list of boxes to cover the field
+         * @param dm the distribution mapping for load balancing with MPI
+         * @param ncomp the number of components of the field (all with the same staggering)
+         * @param ngrow the number of guard (ghost, halo) cells
+         * @param initial_value the optional initial value
+         * @param remake follow the default domain decomposition of the simulation
+         * @param redistribute_on_remake redistribute on @see amrex::AmrCore::RemakeLevel
+         * @return pointer to newly allocated MultiFab
+         */
+        template<typename T>
+        amrex::MultiFab*
```

Comment:
Do we actually need to return the MultiFab here?
I think that having this function be `void` might be cleaner.